feat(new lint): new lint `manual_retain`

kyoto7250 opened the pull request:
close #8097 

This PR is  a new  lint implementation.
This lint checks if the `retain` method is available.

Thank you in advance.

changelog: add new ``[`manual_retain`]`` lint

rust-highfive commented:
r? @llogiq

(rust-highfive has picked a reviewer for you, use r? to override)

llogiq commented on clippy_lints/src/use_retain.rs:
```diff
@@ -0,0 +1,233 @@
+    /// ### What it does
+    /// Checks for code to be replaced by `.retain()`.
+    /// ### Why is this bad?
+    /// `.retain()` is simpler.
```
```suggestion
    /// `.retain()` is simpler and avoids needless allocation.
```

llogiq commented on clippy_lints/src/use_retain.rs:
````diff
@@ -0,0 +1,233 @@
+    /// ```
+    #[clippy::version = "1.63.0"]
+    pub USE_RETAIN,
+    style,
````
I don't think this is a question of `style`. I'd place it in either `performance` (because it helps avoid allocating a new container) or `complexity` (because it simplifies the code) with a small preference to the former.

llogiq commented on clippy_lints/src/use_retain.rs:
```diff
@@ -0,0 +1,233 @@
+    left_expr: &hir::Expr<'_>,
+    target_expr: &hir::Expr<'_>,
+) {
+    if_chain! {
```
Feel free to also use let chains here.

llogiq commented on clippy_lints/src/use_retain.rs:
```diff
@@ -0,0 +1,233 @@
+    left_expr: &hir::Expr<'_>,
+    target_expr: &hir::Expr<'_>,
+) {
+    if_chain! {
```
need I say more? :smile:

llogiq commented on clippy_lints/src/use_retain.rs:
```diff
@@ -0,0 +1,233 @@
+}
+
+fn match_acceptable_def_path(cx: &LateContext<'_>, collect_def_id: DefId) -> bool {
+    return ACCEPTABLE_METHODS
+        .iter()
+        .any(|&method| match_def_path(cx, collect_def_id, method));
```
```suggestion
    ACCEPTABLE_METHODS
        .iter()
        .any(|&method| match_def_path(cx, collect_def_id, method))
```

Why wasn't this caught by `needless-return`!?

kyoto7250 replied:
I made the bug ticket,  I'll check when I have time.

https://github.com/rust-lang/rust-clippy/issues/8978

llogiq commented on clippy_lints/src/use_retain.rs:
```diff
@@ -0,0 +1,233 @@
+
+fn match_acceptable_type(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+    let expr_ty = cx.typeck_results().expr_ty(expr).peel_refs();
+    return ACCEPTABLE_TYPES
+        .iter()
+        .any(|&ty| is_type_diagnostic_item(cx, expr_ty, ty));
```
```suggestion
    ACCEPTABLE_TYPES
        .iter()
        .any(|&ty| is_type_diagnostic_item(cx, expr_ty, ty))
```
same here

llogiq commented on tests/ui/use_retain.rs:
```diff
@@ -0,0 +1,190 @@
+    // https://github.com/rust-lang/rust/issues/71503
+    let mut heap = BinaryHeap::from([1, 2, 3]);
+    heap = heap.into_iter().filter(|x| x % 2 == 0).collect();
+    heap = heap.iter().filter(|&x| x % 2 == 0).copied().collect();
```
Perhaps also one case with `.cloned()` instead of `.copied()` (allowing the respective lint of course)?

llogiq reviewed (commented):
This looks pretty good. I only have some small-ish nits regarding code style and one suggestion for extending the test. With those out of the way this is good to merge.

mikerite commented:
The name does follow the [naming guidelines](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints). It should probably be called something like 'manual_retain' or 'slow retain'.

--

I ran this on (more or less) the top 1000 crates of crates.io.

I got one warning and the fix worked so that is looking good.

```
---> annotate-snippets-0.9.1/src/display_list/from_snippet.rs:344:9   
warning: this expression can be written more simply using `.retain()` 
   --> src/display_list/from_snippet.rs:344:9
    |
344 | /         annotations = annotations
345 | |             .into_iter()
346 | |             .filter(|annotation| {
347 | |                 let body_idx = idx + annotation_line_count;
...   |
507 | |             })
508 | |             .collect();
    | |______________________^
    |
    = note: requested on the command line with `-W clippy::use-retain`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_retain
help: consider calling `.retain()` instead
    |
344 ~         annotations.retain(|annotation| {
345 +                 let body_idx = idx + annotation_line_count;
346 +                 let annotation_type = match annotation.annotation_type {
347 +                     snippet::AnnotationType::Error => DisplayAnnotationType::None,
348 +                     snippet::AnnotationType::Warning => DisplayAnnotationType::None,
349 +                     _ => DisplayAnnotationType::from(annotation.annotation_type),
  ...
annotate-snippets-0.9.1 - fix succeeded
```

kyoto7250 commented:
@llogiq 

I added some commits for your suggestions and I changed the lint name to manual_retain. 

In addition to,  I noticed that this lint needs to check msrv, so I added [the commit](https://github.com/rust-lang/rust-clippy/pull/8972/commits/50af5747a8ee6807b32a93e41e15e7492ae5908f).

Please let me know if we need `git squash` :)

bors commented:
:umbrella: The latest upstream changes (presumably #8964) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #8989) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #9031) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

llogiq commented:
Sorry, I had a lot to do this week, so I didn't get to review stuff. This looks good and can be merged after a rebase.

kyoto7250 commented:
>   This looks good and can be merged after a rebase.

I rebased latest master.

Do I need a git squash?

> Sorry, I had a lot to do this week, so I didn't get to review stuff.

I don't mind.

Thanks again for your reviews  :)

bors commented:
:pushpin: Commit 676af4530f08aeb2aaf561b4cc64e046e5f3e80f has been approved by `llogiq`

<!-- @bors r=llogiq 676af4530f08aeb2aaf561b4cc64e046e5f3e80f -->
<!-- homu: {"type":"Approved","sha":"676af4530f08aeb2aaf561b4cc64e046e5f3e80f","approver":"llogiq"} -->

bors commented:
:hourglass: Testing commit 676af4530f08aeb2aaf561b4cc64e046e5f3e80f with merge eaa03ea911662b282d3c7bc7c3a29f9fb370b933...
<!-- homu: {"type":"BuildStarted","head_sha":"676af4530f08aeb2aaf561b4cc64e046e5f3e80f","merge_sha":"eaa03ea911662b282d3c7bc7c3a29f9fb370b933"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/7074338111?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/7074297979?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/7074577803?check_suite_focus=true)
Approved by: llogiq
Pushing eaa03ea911662b282d3c7bc7c3a29f9fb370b933 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"llogiq","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/7074577803?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/7074338111?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/7074297979?check_suite_focus=true"},"merge_sha":"eaa03ea911662b282d3c7bc7c3a29f9fb370b933"} -->
